sql/rowexec/rel.go

```diff
@@ -644,50 +644,66 @@ func (b *BaseBuilder) buildIndexedTableAccess(ctx *sql.Context, n *plan.IndexedT
 	return sql.NewSpanIter(span, sql.NewTableRowIter(ctx, n.Table, partIter)), nil
 }
 
-func (b *BaseBuilder) buildUnion(ctx *sql.Context, u *plan.Union, row sql.Row) (sql.RowIter, error) {
-	span, ctx := ctx.Span("plan.Union")
-	var iter sql.RowIter
+func (b *BaseBuilder) buildUnion(ctx *sql.Context, s *plan.SetOp, row sql.Row) (sql.RowIter, error) {
+	span, ctx := ctx.Span("plan.SetOp")
+	var iter, iter2 sql.RowIter
 	var err error
-	iter, err = b.buildNodeExec(ctx, u.Left(), row)
-
+	iter, err = b.buildNodeExec(ctx, s.Left(), row)
+	if err != nil {
+		span.End()
+		return nil, err
+	}
+	iter2, err = b.buildNodeExec(ctx, s.Right(), row)
 	if err != nil {
 		span.End()
 		return nil, err
 	}
-	iter = &unionIter{
-		cur: iter,
-		nextIter: func(ctx *sql.Context) (sql.RowIter, error) {
-			return b.buildNodeExec(ctx, u.Right(), row)
-		},
+	switch s.SetOpType {
+	case plan.UnionType:
+		iter = &unionIter{
+			cur:  iter,
+			next: iter2,
+		}
+	case plan.IntersectType:
+		iter = &intersectIter{
+			lIter: iter,
+			rIter: iter2,
+		}
+	case plan.ExceptType:
+		iter = &exceptIter{
+			lIter: iter,
+			rIter: iter2,
+		}
 	}
-	if u.Distinct {
+
+	if s.Distinct {
```

Comment:
Be careful: This works for `UNION` and `INTERSECT`, but not `EXCEPT`. `DISTINCT(TABLE a EXCEPT ALL TABLE b)` is not the same as `TABLE a EXCEPT DISTINCT TABLE b`.

`EXCEPT ALL` subtracts occurrences in the right table from the left table. So a value that appears three times on the left and once on the right would appear twice in the output.

`EXCEPT DISTINCT` computes the set difference: if a row appears on the right, it will never appear in the output, regardless of how many times it appears on the left. 

Thus, the following have different results:

```
mysql> SELECT DISTINCT * FROM (VALUES ROW(1), ROW(1), ROW(1) EXCEPT ALL VALUES ROW(1)) sub_expr;
+----------+
| column_0 |
+----------+
|        1 |
+----------+
1 row in set (0.00 sec)

mysql> VALUES ROW(1), ROW(1), ROW(1) EXCEPT DISTINCT VALUES ROW(1);
Empty set (0.00 sec)
```

It should be correct to insert the DistinctIters into the *inputs* to the Except:

```
mysql> SELECT DISTINCT * FROM (VALUES ROW(1), ROW(1), ROW(1)) sub_expr_l EXCEPT ALL SELECT DISTINCT * FROM (VALUES ROW(1)) sub_expr_r;
Empty set (0.00 sec)
```

Reply:
Good catch, I fixed the logical error.

However, we don't actually parse `(VALUES ROW(1), ROW(1), ...)` here and in many other places.
I wrote a test with the same values in tables. I think a fix for parsing `values_statement` would go in a separate PR.